### PR TITLE
Add airbyte-ci command: migrate-to-manifest-only

### PR DESCRIPTION
**Pull request by ChristoGrab**
## What

This PR adds a new airbyte-ci command: `migrate-to-manifest-only`. This command converts a valid low-code connector to manifest-only. It does not accept any additional arguments.

Resolves [#8849](https://github.com/airbytehq/airbyte-internal-issues/issues/8849)

## How

The command is broken up into three distinct steps:

1. First, we validate the following on a given connector:

- The connector is low-code
- The connector is not already tagged as `manifest-only`
- Only the following files exist in the connector's source folder: `manifest.yaml`, `run.py`, `__init__.py` and `source.py`
- The connector is a subclass of `YamlDeclarativeSource`
- The connector does not override the `streams` method (ie, it is not a hybrid low-code/Python connector)

If any of these checks fail, the migration is safely skipped

2. If the connector is a valid candidate, the second step strips it down to only essential files:

- Moves `manifest.yaml` to the root level of the directory
- Moves any non-inline specs to `manifest.yaml`
- Deletes all non-essential files and folders (everything but `manifest.yaml`, `metadata.yaml`, `icon.svg`, and any valid acceptance/integration test files)

3. The third step updates the remaining files to complete the migration:
- Points the connector's acceptance test config to the root-level manifest
- Adds the `language:manifest-only` tag to the connector's metadata and points to the latest `source-declarative-manifest` base image
- Generates an updated `README.md` file

If a failure occurs during the actual migration, the command will revert any changes to the connector's directory.

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/commands.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import asyncclick as click
+from pipelines.airbyte_ci.connectors.context import ConnectorContext
+from pipelines.airbyte_ci.connectors.pipeline import run_connectors_pipelines
+from pipelines.airbyte_ci.connectors.strip.pipeline import run_connectors_strip_pipeline
+from pipelines.cli.dagger_pipeline_command import DaggerPipelineCommand
+
+@click.command(
+    cls=DaggerPipelineCommand,
+    short_help="Migrate low-code connector to manifest-only"
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/commands.py:13]:
```suggestion
    short_help="Migrate a low-code connector to manifest-only"
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/commands.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    ctx.obj["concurrency"],
+    ctx.obj["dagger_logs_path"],
+    ctx.obj["execute_timeout"],
+    ctx.obj["git_branch"],
+    ctx.obj["git_revision"],
+    ctx.obj["diffed_branch"],
+    ctx.obj["is_local"],
+    ctx.obj["ci_context"],
+    ctx.obj["git_repo_url"],
+
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/commands.py:48]:
```suggestion
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+      manifest_file = manifest_file.rename(self.context.connector.code_directory / "manifest.yaml")
+
+      if manifest_file not in self.context.connector.code_directory.iterdir():
+        return StepResult(
+          step=self,
+          status=StepStatus.FAILURE,
+          stdout="Failed to move manifest.yaml to the root level of the directory."
+        )
+      
+      # We don't want to delete the source_<name> folder
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py:116]: We actually would very much love to delete the folder. We don't need it for sure. We should probably not delete `unit_tests/integration`, but then again we should check if it exists and NOT proceed.

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            step=self,
+            status=StepStatus.FAILURE,
+            stdout=f"Failed to delete {file.name}"
+          )
+
+      # 3. Grab the cdk tag from metadata.yaml and update it
+      metadata_file = self.context.connector.code_directory / "metadata.yaml"
+      with open(metadata_file, "r") as file:
+        metadata = yaml.safe_load(file)
+        tags = metadata['data']['tags']
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py:139]: We should add a `langage:manifest-only` tag, not a cdk tag. I don't know if we should delete the `cdk:low-code` and `language:low-code` though. /cc @bnchrch

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        for i, tag in enumerate(tags):
+          if tag == "cdk:low-code":
+            tags[i] = "cdk:manifest-only"
+
+      # Write the changes to metadata.yaml
+      with open(metadata_file, "w") as file:
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py:145]: We'd have to replace yaml.load and yaml.dump with `ruamel.YAML`. There are a few examples lying around. The difference is that vanilla yaml will screw up your keys ordering, and your metadata file won't retain the original key order.

**Review by natikgadzhi — COMMENTED** [review]
Good start!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py (hunk 3 above)
Comment (ChristoGrab) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py]: Gotcha! I removed the folder from the "don't delete" list and will work on adding the check for unit_tests/integration

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py (hunk 4 above)
Comment (ChristoGrab) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py]: Haha I knew that too, my Friday afternoon brain musta kicked in 🫠 Switched to appending the language tag for now and can add the logic to delete the others if needed

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py (hunk 5 above)
Comment (ChristoGrab) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/strip/pipeline.py]: Done!

**Review by natikgadzhi — APPROVED** [review]
Please update the versions, and merge this, I don't want the code to sit for too long. 

Make a follow-up PR (or a commit in this one) that parses the manifest and detects if `$parameters` are there and stops. 

Other than that, we can run the wave.

**Comment by natikgadzhi** [comment]
@ChristoGrab new problems: source-activecampaign publishing fails in a funny way:

- gh run: https://github.com/airbytehq/airbyte/actions/runs/10223174041/job/28289079809
- report: https://ff9a4c2d768d00641202104c799ad0b0e60577efb4c7e966b4b3ff3-apidata.googleusercontent.com/download/storage/v1/b/airbyte-ci-reports-multi/o/airbyte-ci%2Fconnectors%2Fpublish%2Fmaster%2Fmaster%2F1722640435%2F3be2e97d8743ab3059be00625342ce672e43e898%2Fsource-activecampaign%2F0.2.0%2Foutput.html?jk=AYBlUPCRQofAIVI5Oxi3D45VdEgGuS7LCu5UrKWXRSDOPth8rHfDYaCv-C7RI3adKyaGWhd9V7Cy9emkFtr2IMp9AkqjHfcaRVVqZqqTARSiwEsIHT8chbmyO7YkqocZB-2LOxqV8cKVDCa10ouRqx-YOeCMsZLbuMAZeb4a7OZ70iUvJGl0hCnZN9W3HwPxSSR9jaR8nVinuE5h4bBy8PN9ZYLmVRwbUDYRozGpp3wTiiz6LDQ_VdN62eSnewiBoCSzrXyYzHEMr4n5r9FPX9BWlR6zpmdLiCykG5Jpp8xe3lNxofyt53OvDA4r90qLfh6tvnipKX3jPRMtTU0lDU9w3cS2t_NL6_151IIg9k67TnN_u_I-Q99-TQcTaJ8zNhVe0o8cB5ACYu7YnspDZ8dcjV77hTsNsAVwYw4IsnC5KY1fQ6zLSU-AOkz5_LmME8hfMDq79-BMCjBH3-Ua61nJYJbtqsxhmQJ8HhNUbCwckm2TRvwvSdl4WZYl5DPHgjxon85foU9chZvTHHY_sNzzW_WE1p1cp9mHt4LZ6iZ3gtt0YJLKMdt-Ca8jgBkhDfa-wp9ojg8zJ0ifIso1wYp4y-pPdzRIrIDwqHJ1sjCLLyD1AmYdotcUldGHura1D6J023v2FPRI7vsNqMCK-XHgzzmbuRYi8RTaONynqZfpEnevHSHHElkgbA38_NyRiuYYEOMkLijX_cKew2JdeQDl1rn1TVHJ-r8A8y0zXhHlgaXnDxJIQ-hVALl0ikR2lCuzIsB3A3WuMZRpP5UrBPtCSsArg6TcDZG75aInbOTeGkbnSI24PP6RPfRpJbAbpOX6LqTrUJ14RsTFt5-nJuxcOn2hlpVEzhr8cGz6fnuFFoF8KCScv0HpW3MNjmFT55xeeR5KdzFjPR3vhBj4AQZoe31Gdfv4eAKQeBJR5J7AmA7mNn2H-dGRKLwdEZYNWv_cPNSFmqm5lMD5irh9fXNZW0oYk5X6LaVvMZ6wQcooAi67ikBm7CtwOMyH1R78UfbfuMhhINjQAMWKZQBVTY72u99lGYZK8El7T920RPiiLlK-EI0Ux8QjURxV-tfXmN8r7IqLg7Xy7CRQauzbkpDlHgV0JBZL5LSPGACK3-keQTELZZdkHhASyKMqhdqbhKYgOoNPlvFfG6di4IIuEFzsamNZuqCJGa7-PgHEFReiUZd60b4qivwD0q7xd_v0GqwN9km04qoGBELGaGBxgsZyLso-R0LrwioXHHyLeArm34cXiXxENmUXs5RuKEPlp16Nbl_08nQuDPwkBl5_YNC5HniuGuu5CfNIC8Sm9jWjHf4asFRVtzScDkupxjPTYvixQszZ2zJZFSfbie4F76zOn9JhRHJ_Mf_Jlg&isca=1

Basically, it says that if it's a pypi connector, it must have python language tag. The answer to this is to remove disable the pypi entry in metadata if it exists when converting.

Long-term, this creates a little problem: there are packages for connectors on pypi that we are going to pull from pypi and start publishing as manifest-only. We should have a way to tell pypi users about it.

**Comment by ChristoGrab** [comment]
@natikgadzhi Hey, just coming back from the long weekend! Heard on the pypi tagging, I'll add logic to disable the tag and get this merged 👍

File: airbyte-ci/connectors/pipelines/README.md [hunk 6]
```diff
@@ -769,17 +771,33 @@ You can pass multiple `--poetry-package-path` options to run poe tasks.
 E.G.: running Poe tasks on the modified internal packages of the current branch:
 `airbyte-ci test --modified`
 
+### <a id="migrate-to-manifest-only-command"></a>`migrate-to-manifest-only` command
+
+This command migrates valid connectors to the `manifest-only` format. It contains three steps:
+
+1. Check -> Validates whether a connector is a candidate for the migration. If not, the operation will be skipped.
+2. Strip -> Strips out all unneccessary files/folders, leaving only the root-level manifest, metadata, icon, and acceptance/integration test files
+3. Update -> Updates the connector's metadata and test-config, and generates a templated README
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/README.md:780]:
```suggestion
This command migrates valid connectors to the `manifest-only` format. It contains two steps:

1. Check: Validates whether a connector is a candidate for the migration. If not, the operation will be skipped.
2. Migrate: Strips out all unneccessary files/folders, leaving only the root-level manifest, metadata, icon, and acceptance/integration test files. Unwraps the manifest (references and `$parameters`) so it's compatible with Connector Builder.
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/migrate_to_manifest_only/utils.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+def remove_parameters_from_manifest(d: Mapping[str, Any] | List) -> Mapping[str, Any] | List:
+    """
+    Takes a dictionary (or a list) of keys and removes all instances of the key "$parameters" from it.
+    """
+    if isinstance(d, dict):
+        return {k: remove_parameters_from_manifest(v) for k, v in d.items() if k != "$parameters"}
+    elif isinstance(d, list):
+        return [remove_parameters_from_manifest(item) for item in d]
+    else:
+        return d
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/migrate_to_manifest_only/utils.py:90]:
```suggestion
        return d
        
```

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/migrate_to_manifest_only/utils.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+import subprocess
+from pathlib import Path
+
+from typing import Union, Dict, List, Mapping, Any
+
+import jinja2
+import requests
+
+## HELPER FUNCTIONS
```
Comment (natikgadzhi) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/migrate_to_manifest_only/utils.py:13]:
```suggestion
```

**Comment by natikgadzhi** [comment]
Using this to migrate source-hibob and source-northpass-lms that got merged yesterday ;)
